### DISCUSSION TITLE
callback's wasm_val_vec_t* results contain invalid value kind

### DISCUSSION THREAD

**Discussion by mean-ui-thread** [wasmerio/wasmer#3009] [General]
### Summary

instead of doing this:

```C
wasm_trap_t* host_func_callback(const wasm_val_vec_t* args, wasm_val_vec_t* results) {
    printf("Calling back...\n> ");

    wasm_val_t val = WASM_I32_VAL(42);
    wasm_val_copy(&results->data[0], &val);

    wasm_val_delete(&val);

    return NULL;
}
```
I'd like to be able to do this to avoid allocating wasm_val_t for performance reasons. This is more efficient:
```C
wasm_trap_t* host_func_callback(const wasm_val_vec_t* args, wasm_val_vec_t* results) {
    printf("Calling back...\n> ");

    results->data[0].kind = WASM_I32;
    results->data[0].of.i32 = 42;

    return NULL;
}
```
But here's the question: why do I need to set `results->data[0].kind` ? If I don't, I get an error like this:

```
Dynamic function returned wrong signature. Expected [I32] but got [I64]
```
Obviously, wasmer knows what the value kind is, so why isn't it setting the valkind correctly in the `wasm_val_vec_t* results` vector for us? If that's done, then I'd be able to do just this:

```C
wasm_trap_t* host_func_callback(const wasm_val_vec_t* args, wasm_val_vec_t* results) {
    printf("Calling back...\n> ");

    results->data[0].of.i32 = 42;

    return NULL;
}
```
which would be the best/simplest/minimalistic code.

### Additional details

I'm on 2.3.0 on Linux Desktop.

**Comment by syrusakbary** [comment]
For future reference, the error is coming from here: https://github.com/wasmerio/wasmer/blob/master/lib/api/src/sys/externals/function.rs#L818-L824

**Comment by syrusakbary** [comment]
Thanks for creating the issue @mean-ui-thread. I've converted this into a discussion so other people can also chime in into what are the best ways to improve DX (Developer Experience) on it (if any)
